### DISCUSSION TITLE
Confused about the bank plan and upgrading an empty bank

### DISCUSSION THREAD

**Discussion by EnsignR** [hack-technicolor/hack-technicolor#211] [Q&A] [answered]
I followed the guide to root a DJA0231, already on the latest TYPE 2 (strategy C) firmware, 20.3.c.0389-MR20-RA.
Using the guide this takes you straight from the Rooting Type 2 page to Bank Planning section in the Post-Root Procedures.

Prior to pasting in the commands to achieve the "optimal bank plan" my banktable is described as follows.

```
/proc/banktable/active
bank_2

/proc/banktable/booted
bank_2

/proc/banktable/inactive
bank_1

/proc/banktable/bootedoid
61884d0a588f1c7511457667

/proc/banktable/notbooted
bank_1

/proc/banktable/notbootedoid
60f5197344321408f315ab5f

/proc/banktable/activeversion
20.3.c.0389-xxxx

/proc/banktable/passiveversion
20.3.c.0329-xxxx
```

This to my untrained (though CS degree backed) eye, the device had the **two latest firmware** installed on each of it's banks. The _activeversion_ begin the latest/current firmware, 20.3.c.0389, marked as both _active_ and _booted_ on **bank_2** and the previous, _passiveversion_, firmware on the _inactive_/_notbooted_ **bank_1**.

After running the supplied commands the banktable is as follows
```
/proc/banktable/active
bank_1

/proc/banktable/booted
bank_2

/proc/banktable/inactive
bank_2

/proc/banktable/bootedoid
61884d0a588f1c7511457667

/proc/banktable/notbooted
bank_1

/proc/banktable/notbootedoid
Unknown
/proc/banktable/activeversion
Unknown
/proc/banktable/passiveversion
20.3.c.0389-xxxx
```

As per the above and corroborated by the text on the page **bank_1** is now empty (though still _active_) and **bank_2** has been _booted_, using the latest firmware (which was already installed) after trying and failing to boot the empty **bank_1** partition.

Next I skipped setting **SSH** permanently to port 22 because it already was, and I reset the **root password**, the next step according to the guide is **Hardening Root Access**. I do note that in the smaller text in the break out box it does suggest upgrading via **Safe Firmware Upgrade** page prior to **SSH** section. TBH I find this a bit confusing as to which is the correct path to follow. Noting that my **bank_1** is currently **empty**, and the whole point of the bank planning is to keep a fresh, always bootable firmware (i.e. my **bank_2**) I believe the better choice is to add another firmware on **bank_1**, which would mean heading to the **SFU** page.

However I initially went on to **Hardening Root Access** (most of the commands here failed FWIW), before following onto the _next_ page, **Unlock Functionality**, where I realised I probably shouldn't be modifying the **only firmware currently on the device**.

This is where I really get confused. On the **Safe Firmware Upgrade** page it states:

> On which bank will the new firmware be flashed?
> - The booted one, the currently installed firmware will be overwritten. Make sure it is listed in our [Repository](https://hack-technicolor.readthedocs.io/en/stable/Repository/) before overwriting!

As above I'm currently **booting** to **bank_2**, which is the **only firmware currently on the device**. 

My understanding is what I really need is to put another copy of the same firmware onto **bank_1** which I can than play around with to my heart's content. Except I can't find any instructions on how exactly I am supposed to do that.

Sorry for the long post, but it seems I've followed the guide but got to a point where I don't know how to safely proceed with the next step. Whilst I might be able to figure it out myself I think it's worth asking the question, both to save me the trouble, and to make sure that I, and anyone else who gets to the same position, knows the correct way forward.

Thanks.

**Comment by LuKePicci** [comment, marked as answer]
Despite being counterintuitive (which I confirm) the guide says the truth about bank planning and other "safety-related" aspects.

The Homeware platform relys on dual-bank firmware to provide fault-tolerance during firmware upgrade. However our requisites are different. Please, read further about [different methods of flashing firmware](https://hack-technicolor.readthedocs.io/en/latest/Resources/#different-methods-of-flashing-firmwares).

When you get a regular upgrade the other bank is used to store the new firmware and then a config migration occurs. That config migration mechanism does only support what is inside native Homeware feature set. This means whatever stuff we customize would be ignored by such migration. For this reason, the only safe way of upgrading firmware when you have custom stuff in it is to start from fresh state (+ root access, for convenience). That is what the Safe firmware upgrade guide is proposing.

The suggested bank planning comes from different considerations. TFTP recovery is only capable of writing into bank_1. If you get in (serious) troubles with your **booted** bank and your booted firmware was into bank_1 your chances to recover start decreasing quite quickly because of the bank_1 overlay is f.'d up and whatever firmware you flash by TFTP will fail to boot properly because of that problematic bank_1 overlay contents. The short thing is that the following are sufficient conditions to never get bricked:
- keep bank_1 empty
- keep bank_1 overlay clean/empty
- keep it set to always try booting bank_1 first (i.e. _active_)

> I do note that in the smaller text in the break out box it does suggest upgrading via Safe Firmware Upgrade page prior to SSH section.

OFC you don't need to upgrade if you are on latest firmware already, that note is for people who are on older releases. The reason for upgrading at that point is that any further configuration would be cleaned-up by the Safe Upgrade and must be done afterwards.

**Comment by EnsignR** [comment]
Thanks for your reply. TBQH still a bit confused but I think I get I (mostly) get it now.

Since this is kinda related to bank planning as well I'll tag my follow up question here too.

The [Recovery page](https://hack-technicolor.readthedocs.io/en/latest/Recovery/) mentions

> Fix it manually if you know what was your mistake, otherwise restore /overlay/bank_2 folder contents from a previous overlay backup. If you didn't get any (shame on you)

However if (at least AFAIK) in following the guide the only place a backup is done is on [Post Root Procedures
](https://hack-technicolor.readthedocs.io/en/latest/Hacking/PostRoot/) with following command
`cp -rf /overlay/$(cat /proc/banktable/booted) /tmp/bank_overlay_backup`

The problem is that this backup is removed on reboot when it is forced in a few commands below it.

In other words following the guide you end up with no accessible overlay backup if you screw things up; or at least I'm not getting one.

Should the backup be saved to a directory that doesn't get re-populated at reboot?
Or should there really be an extra step where you create the backup in SSH and then copy it to your PC with SCP?

Thanks.

**Comment by LuKePicci** [comment]
> Or should there really be an extra step where you create the backup in SSH and then copy it to your PC with SCP?

This 😁

**Reply by EnsignR** [reply]
OK. So based on this and the information on the Resources page I've dusted off my bash skills and created a section at the end of the Post-Root page (see my [fork](https://github.com/EnsignR/hack-technicolor)).

To make it easier this requires the use of a USB flash drive, but also refers reader to Resources page for more info / alternative instructions. My initial thought was to use SCP to move each dump after it was created, but my device reset writing mtd1 to /tmp; I assume because the filesystem was full.

One thing that puzzled me is that in the Resources page is states the the mtds for bank1 and 2 won't have the same checksum without changing the first 4 bytes. I've found this not to be the case (I also use cmp, because why not?).

```
dev:    size   erasesize  name			md5sum/cmp matches
mtd0: 20000000 00040000 "brcmnand.0"	no
mtd1: 091a0000 00040000 "rootfs"		yes
mtd2: 0c840000 00040000 "rootfs_data"	no
mtd3: 095c0000 00040000 "bank_1"		yes
mtd4: 095c0000 00040000 "bank_2"		yes
mtd5: 00040000 00040000 "eripv2"		yes
mtd6: 00080000 00040000 "rawstorage"	yes
```

I assume that NAND.0 is where BOOT0 is located and they're masking is somehow for ... well you know,  and the rootfs_data is different because bash history is begin written after the dd and before the md5sum or file comparison.

So, are the above assumptions correct?
Also if you've had a look at the update in my [fork](https://github.com/EnsignR/hack-technicolor), would you do anything different.

Peas.
